DeDir: Add new pallet `Decentralized Directory`

vatsa287 opened the pull request:
A `v1` implementation of DeDir - Decentralised Directory.

- [x] Basic Functionalities - `create-registry`, `create-registry-entry`, `registry-entry-state-change`.
- [x] Change states to be in `enum` format. Custom states will be from `attributes`.
- [x] Identifier creation/management.
- [x] Delegation/ Registry-Ownership management.


Abandoned. Do not merge.

vatsa287 committed:
DeDir: Add new pallet Decentralized Directory

Signed-off-by: Shreevatsa N <vatsa@dhiway.com>

amarts reviewed (dismissed):
LGTM

Is there any demo script planned to be implemented for the same?

vatsa287 commented:
> LGTM
> 
> Is there any demo script planned to be implemented for the same?

@amarts 
Sure it will be implemented.
It would be great if you can merge [this PR](https://github.com/dhiway/cord.js/pull/231), so I can push on top of this.

smohan-dw commented on pallets/dedir/src/lib.rs:
```diff
@@ -0,0 +1,456 @@
+		#[pallet::weight({0})]
+		pub fn create_registry(
+			origin: OriginFor<T>,
+			registry_id: RegistryIdOf<T>,
```
Registry ID is an identifier that should be returned after a successful operation.

vatsa287 replied:
As per earlier conversation `identifier` creation was passed onto app side. 
Would move the logic here if required.

smohan-dw commented on pallets/dedir/src/lib.rs:
```diff
@@ -0,0 +1,456 @@
+		pub fn create_registry(
+			origin: OriginFor<T>,
+			registry_id: RegistryIdOf<T>,
+			attrs: Vec<(RegistryKeyIdOf<T>, RegistrySupportedTypeOf)>,
```
The registry entry attributes should also help to identify the lookup fields that should be used for creating entry identifiers

vatsa287 replied:
With the discussion with @amarts  previously if the identifier is available then the registry_entry attributes would be fetch able.

smohan-dw commented on pallets/dedir/src/lib.rs:
```diff
@@ -0,0 +1,456 @@
+			registry_id: RegistryIdOf<T>,
+			attrs: Vec<(RegistryKeyIdOf<T>, RegistrySupportedTypeOf)>,
+			// TODO: Use Runtime Constant for upper bound
+			additional_states: Option<BoundedVec<BoundedVec<u8, ConstU32<32>>, ConstU32<10>>>,
```
why can't this be enums? also why can't this be part of attrs?

vatsa287 replied:
Ack.
As per newer design the states supported shall not be dynamic as implemented above but fixed to `Active`, `Revoked`, `Draft` which will become enum type.
So any custom state can be part of attrs itself.

smohan-dw commented on pallets/dedir/src/lib.rs:
```diff
@@ -0,0 +1,456 @@
+				}
+			}
+
+			Registries::<T>::insert(&registry_id, registry);
```
registry ownership?

vatsa287 replied:
Delegation/Ownership is not implemented in this version. Will do with next commit.

smohan-dw commented on pallets/dedir/src/lib.rs:
```diff
@@ -0,0 +1,456 @@
+		pub fn create_registry_entry(
+			origin: OriginFor<T>,
+			registry_id: RegistryIdOf<T>,
+			registry_entry_id: RegistryEntryIdOf<T>,
```
again registry_entry_id is the result of the operation

smohan-dw commented on pallets/dedir/src/lib.rs:
```diff
@@ -0,0 +1,456 @@
+			digest: EntryHashOf<T>,
+			attrs: Option<Vec<(RegistryKeyIdOf<T>, Data)>>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
```
need to use this to check the registry ownership or delegation

vatsa287 committed:
Identifier management & Attribute State as Enums

- Identifier management: Validation of ss58 and ident type on chain
- State as enums: Make states as enums, dynamic states can be through attributes
- Address review comments

Signed-off-by: Shreevatsa N <vatsa@dhiway.com>

vatsa287 committed:
Delegation & Identifier Management Updates

Signed-off-by: Shreevatsa N <vatsa@dhiway.com>

amarts commented:
done with #504
